src/jit/codegenxarch.cpp

```diff
@@ -176,23 +176,86 @@ void                CodeGen::genEmitGSCookieCheck(bool pushReg)
-        gcInfo.gcRegGCrefSetCur |= RBM_INTRET;    
+    //
+    // For Amd64 System V, a two-register-returned struct could be returned in RAX and RDX
+    // In such case make sure that the correct GC-ness of RDX is reported as well, so
+    // a GC object pointed by RDX will not be collected.
+    if (!pushReg)
+    {
+
+#ifdef FEATURE_UNIX_AMD64_STRUCT_PASSING
+        if (compiler->compMethodReturnsMultiRegRetType())
+        {
+            // Get the return tye of the struct.
+            SYSTEMV_AMD64_CORINFO_STRUCT_REG_PASSING_DESCRIPTOR retStructDesc;
+            compiler->eeGetSystemVAmd64PassStructInRegisterDescriptor(compiler->info.compMethodInfo->args.retTypeClass, &retStructDesc);
+
+            assert(retStructDesc.passedInRegisters);
+
+            // In case the return type is a two-register-return, the native return type should be a struct
+            assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+            // In case of a single register returned struct, the return type have been normalized.
+            // In such case the execution will not be here 
+            // (the if (compiler->compMethodReturnsMultiRegRetType()) will evaluate to false).
+            //
+            // The case of return in an implicit RetBuf and the RetBuff address is returned (on Windows or System V)
+            // in REG_INTRET is handled below (in the else part).
+            assert(retStructDesc.eightByteCount == CLR_SYSTEMV_MAX_EIGHTBYTES_COUNT_TO_PASS_IN_REGISTERS);
+
+            unsigned __int8 offset0 = 0;
+            unsigned __int8 offset1 = 0;
+
+            var_types type0 = TYP_UNKNOWN;
+            var_types type1 = TYP_UNKNOWN;
+
+            // Set the GC-ness of the struct return registers.
+            getStructTypeOffset(retStructDesc, &type0, &type1, &offset0, &offset1);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET, type0);
+            gcInfo.gcMarkRegPtrVal(REG_INTRET_1, type1);
+        }
+        else
+#endif // FEATURE_UNIX_AMD64_STRUCT_PASSING
+        {
+            if (compiler->compMethodReturnsRetBufAddr())
+            {
+                // This is for returning in an implicit RetBuf.
+                // If the address of the buffer is returned in REG_INTRET, mark the content of INTRET as ByRef.
+
+                // In case the return is in an implicit RetBuf, the native return type should be a struct
+                assert(varTypeIsStruct(compiler->info.compRetNativeType));
+
+                gcInfo.gcMarkRegPtrVal(REG_INTRET, TYP_BYREF);
+            }
+            else
+            {
+                // In case the function is not a two-register-return
+                // and not returning in the implicit RetBuf, which address is returned in REG_INTREG,
+                // the native return type should have been normalized or is TYP_STRUCT.
+                assert((compiler->info.compRetBuffArg != BAD_VAR_NUM) ||
+                       (!varTypeIsStruct(compiler->info.compRetNativeType)));
```

Comment:
I am not sure whether can assert anything meaningful on Windows

On Unix the following two asserts must hold but not necessarily on windows:
assert(compiler->info.compRetBuffArg == BAD_VAR_NUM);
assert(!varTypeIsStruct(compiler->info.compRetNativeType));